### ISSUE TITLE
see if mysql has_table() can support temp tables w/ DESCRIBE (again, unfortunately)

### ISSUE THREAD

**Issue by scastria** [sqlalchemy/sqlalchemy#9058]
### Describe the bug

Version 1.3.13 would use DESCRIBE table_name to check the existence of a table for has_table function.  This technique DOES detect the presence of a TEMPORARY table.  However, with version 1.4.46, the has_table function implementation has changed to use SELECT COUNT(*) FROM information_schema.tables which does NOT detect the presence of a TEMPORARY table.

### To Reproduce

```python
import pandas as pd

df = pd.DataFrame({"num": [1, 2]})
engine = <construct an engine to any MySQL DB>
with engine.begin() as con:
    con.execute("CREATE TEMPORARY TABLE shawn (num int)")
    df.to_sql("shawn", con=con, index=False)
    results = con.execute("SELECT * FROM shawn")
    for row in results:
        print(row)
```

### Error

No error code, but the problem with the above code is that pandas uses sqlalchemy to check for existence of the shawn table and it returns false since the new existence check in sqlalchemy no longer detects TEMPORARY tables.  So pandas creates a new REAL table with the same name.  Then when pandas writes to the table using INSERT, the TEMPORARY table is used since TEMPORARY tables take precedence.  When the connection is closed, the TEMPORARY table automatically gets removed, but the REAL table mistakenly created by pandas gets left behind.

### Versions

- OS: MacOS 13.1
- Python: 3.8
- SQLAlchemy: 1.4.46
- Database: MySQL 8
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pymysql

**Comment by zzzeek** [comment]
@CaselIT has_table appears to return True for PG, SQLite.   didn't we go over all these rules for the new reflection patch in 2.0?

test case

```py
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import inspect
from sqlalchemy import text

engine = create_engine("mysql://scott:tiger@localhost/test", echo=True)
#engine = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
#engine = create_engine("sqlite://", echo=True)

with engine.begin() as con:
    con.execute(text("CREATE TEMPORARY TABLE shawn (num int)"))

    insp = inspect(con)
    assert insp.has_table("shawn")

```

**Comment by zzzeek** [comment]
this was never an intended use case for has_table() and there's not been any test support for this case, otherwise we'd have known about it.   release 1.4 has been out for two years now and it's at version 1.4.46 so I'm concerned it's a little late to make a sudden behavioral change like this.  it requires adding an additional query against innodb_temp_table_info because it's not an option to use DESCRIBE here anymore.

@CaselIT what do you think about adding has_temporary_table(), removing all temp table detection from has_table() for all dialects and doing another rc?  this really shouldn't be an accidental API

**Comment by zzzeek** [comment]
MariaDB actually has *no* information schema for this; their API was removed per https://jira.mariadb.org/browse/MDEV-12459.    MySQL has innodb_temp_table_info, which says nothing about other storage engines.  There's also a table_exists() function but it appears to be vapor on mariadb.

so this is so far not looking easy without reverting entirely to using DESCRIBE.

**Comment by zzzeek** [comment]
OK current reqs for this are defined in 2.0 as 

```py
    @property
    def has_temp_table(self):
        """target dialect supports checking a single temp table name

        unfortunately this is not the same as temp_table_names

        """

        return only_on(["sqlite", "oracle", "postgresql", "mssql"]) + skip_if(
            self._sqlite_file_db
        )


```

why not sqlite file?   for Mysql /mariadb we'd have to use DESCRIBE and just use the cursor directly to avoid the whole error handling cycle

**Comment by zzzeek** [comment]
works fine for sqlite file, will try to re-enable

**Comment by scastria** [comment]
To my knowledge, MySQL’s innodb_temp_table_info doesn’t have the name of the table so you can’t use it to check for existence. DESCRIBE is the only way that I know of and I only know it as I discovered it when researching what changed in sqlalchemy that caused this bug.

**Comment by zzzeek** [comment]
I dont want this backported to 1.4 because it would be a bigger change for 1.4 than it would be for 2.0.    1.4 has been out for two years so if we knew about this use case 18 months ago, it would have been easier to revert.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**revert MySQL to use DESCRIBE for has_table()** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4344

**Comment by CaselIT** [comment]
I would need to look into what test there are. I'm fairly confident this case is tested in v2 but maybe it is disabled using requirements.

Looking at the linked mariadb issue it seems that an api is still missing

**Comment by CaselIT** [comment]
That requirement was added in PS6 in this gerrit https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4160/5..6 by you mike.

Before `temp_table_names` was used for this. Maybe we could restore the behaviour?

**Comment by CaselIT** [comment]
> Before `temp_table_names` was used for this. Maybe we could restore the behaviour?

No, since mysql does not support listing temp tables

**Comment by zzzeek** [comment]
so when we did dc91c7db7ff32243cd2f6fc04f4e3a6d62f7b11b, I searched pretty hard for why we couldn't change to information_schema, the temp table use case was unknown to me at that time (though it obviously works with DESCRIBE).  We typically rely on tests failing to see if something is a regression.   then when we did db08a699489c9b0259579d7ff7fd6bf  and added has_table() tests for temp, MySQL seemed to not support it, as it also doesn't support listing temp tables, so it seemed like we should just exclude it.   The two events did not get connected (how could they).

we usually rely upon the userbase to let us know for things like this but somehow 1.4 went two years without anyone noticing.
